File: materia/cameraAndFotos/cameraAndFotos/ViewController.m
//
//  ViewController.m
//  cameraAndFotos
//
//  Created by Wallace Goncalves de Oliveira on 03/09/15.
//  Copyright (c) 2015 Wallace Goncalves de Oliveira. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UIImagePickerControllerDelegate, UINavigationControllerDelegate> {
    
}

@property (weak, nonatomic) IBOutlet UIImageView *imageView;

@end

@implementation ViewController

- (UIAlertController *)alertSheetController
{
    return [UIAlertController alertControllerWithTitle:@"Escolha a origem"
                                               message:nil
                                        preferredStyle:UIAlertControllerStyleActionSheet];
}

- (UIImagePickerController *)imagePickerWithSourceType:(UIImagePickerControllerSourceType)sourceType
{
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.sourceType = sourceType;
    picker.delegate = self;
    picker.allowsEditing = YES;
    return picker;
}

- (NSString *)titleForSheetWithSourceType:(UIImagePickerControllerSourceType)sourceType
{
    switch (sourceType) {
        case UIImagePickerControllerSourceTypeCamera:
            return @"Camera";
            
        case UIImagePickerControllerSourceTypePhotoLibrary:
            return @"Photo Library";
            
        case UIImagePickerControllerSourceTypeSavedPhotosAlbum:
            return @"Saved Photos Album";
    }
    
    return nil;
}

- (UIAlertAction *)alertActionForSourceType:(UIImagePickerControllerSourceType)sourceType
{
    void (^handler)(UIAlertAction *) = ^(UIAlertAction *action) {
        UIImagePickerController *picker = [self imagePickerWithSourceType:sourceType];
        [self presentViewController:picker animated:YES completion:nil];
    };
    
    NSString *sheetTitle = [self titleForSheetWithSourceType:sourceType];
    
    return [UIAlertAction actionWithTitle:sheetTitle
                                    style:UIAlertActionStyleDefault
                                  handler:handler];
}

- (IBAction)escolherImagem:(id)sender
{
    UIAlertController *alert = [self alertSheetController];
    
    UIImagePickerControllerSourceType sourceTypes[] = {
        UIImagePickerControllerSourceTypeCamera,
        UIImagePickerControllerSourceTypePhotoLibrary,
        UIImagePickerControllerSourceTypeSavedPhotosAlbum
    };
    
    for (int i = 0; i < 3; i++) {
        if ([UIImagePickerController isSourceTypeAvailable:sourceTypes[i]]) {
            UIAlertAction *action = [self alertActionForSourceType:sourceTypes[i]];
            [alert addAction:action];
        }
    }
    
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel"
                                              style:UIAlertActionStyleDefault
                                            handler:nil]];
    
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - View Lifecycle

-(void)viewDidLoad {
    [super viewDidLoad];
    
    // checando se temos camera disponivel
    BOOL temCamFront = [UIImagePickerController
                          isCameraDeviceAvailable:UIImagePickerControllerCameraDeviceFront];
    
    BOOL temCamTras = [UIImagePickerController
                       isCameraDeviceAvailable:UIImagePickerControllerCameraDeviceRear];
    
    NSLog(@"%@", temCamFront ? @"Cam Front" : @"Sem Cam Front");
    NSLog(@"%@", temCamTras ? @"Cam Tras" : @"Sem Cam Tras");
    
    // checando se temos flash
    BOOL temFlashFront = [UIImagePickerController
                          isFlashAvailableForCameraDevice:UIImagePickerControllerCameraDeviceFront];
    
    BOOL temFlashTras = [UIImagePickerController
                         isFlashAvailableForCameraDevice:UIImagePickerControllerCameraDeviceRear];
    
    NSLog(@"%@", temFlashFront ? @"Cam Flash Cam Front" : @"Sem Flash Cam Front");
    NSLog(@"%@", temFlashTras ? @"Cam Flash Cam Tras" : @"Sem Flash Cam Tras");
    
    // checando os tipos de midia
    NSArray *camaraTypes = [UIImagePickerController
                            availableMediaTypesForSourceType:UIImagePickerControllerSourceTypeCamera];
    
    NSArray *libTypes = [UIImagePickerController
                         availableMediaTypesForSourceType:UIImagePickerControllerSourceTypePhotoLibrary];
    
    NSArray *albumTypes = [UIImagePickerController
                           availableMediaTypesForSourceType:UIImagePickerControllerSourceTypeSavedPhotosAlbum];
    
    NSLog(@"%@", camaraTypes);
    NSLog(@"%@", libTypes);
    NSLog(@"%@", albumTypes);
}

#pragma mark - UIImagePickerControllerDelegate

-(void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info{
    NSLog(@"%@", [info allKeys]);
    
    UIImage *image = info[UIImagePickerControllerOriginalImage];
    [_imageView setImage:image];
    
    [picker dismissViewControllerAnimated:YES completion:nil];
}

-(void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker dismissViewControllerAnimated:YES completion:nil];
}

@end

File: desafios/desafioCoreLocation/desafioCoreLocation/ViewController.swift
//
//  ViewController.swift
//  desafioCoreLocation
//
//  Created by Wallace Goncalves de Oliveira on 23/09/15.
//  Copyright © 2015 Wallace Goncalves de Oliveira. All rights reserved.
//

import UIKit
import MapKit
import CoreLocation
import CoreData

class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate, CLLocationManagerDelegate {
    
    var appDelegate: AppDelegate = UIApplication.sharedApplication().delegate as! AppDelegate;
    var context: NSManagedObjectContext?;
    
    var locationManager = CLLocationManager()
    var checkin: Checkin? = nil
    
    //MARK: Outlets
    
    @IBOutlet weak var mapView: MKMapView!
    
    //MARK: Actions
    
    @IBAction func checkin(sender: UIButton) {
        let alert = alertSheetController()
        
        if UIImagePickerController.isSourceTypeAvailable(UIImagePickerControllerSourceType.Camera) {
            let action = alertActionForSourceType(UIImagePickerControllerSourceType.Camera)
            alert.addAction(action)
        }
        
        alert.addAction(UIAlertAction(title: "Cancel", style: UIAlertActionStyle.Default, handler: nil))
        self.presentViewController(alert, animated: true, completion: nil)
    }
    
    //MARK: Helpers
    
    func alertSheetController()->UIAlertController {
        return UIAlertController(title: "Escolha a origem", message: nil, preferredStyle:UIAlertControllerStyle.ActionSheet)
    }
    
    func alertActionForSourceType(sourceType: UIImagePickerControllerSourceType)->UIAlertAction {
        let sheetTitle = titleForSheetWithSourceType(sourceType)
        return UIAlertAction(title: sheetTitle, style: UIAlertActionStyle.Default) { (action: UIAlertAction) -> Void in
            let picker = self.imagePickerWithSourceType(sourceType)
            self.presentViewController(picker, animated: true, completion: nil)
        }
    }
    
    func titleForSheetWithSourceType(sourceType: UIImagePickerControllerSourceType)->String {
        switch sourceType {
            case UIImagePickerControllerSourceType.Camera: return "Camera";
            case UIImagePickerControllerSourceType.PhotoLibrary: return "Photo Library";
            case UIImagePickerControllerSourceType.SavedPhotosAlbum: return "Saved Photos Album";
        }
    }
    
    func imagePickerWithSourceType(sourceType: UIImagePickerControllerSourceType)->UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = sourceType
        picker.delegate = self
        picker.allowsEditing = true
        return picker
    }
    
    func initContextFromAppDelegate() {
        context = appDelegate.managedObjectContext;
    }
    func initLocationManager() {
        locationManager.delegate = self
        locationManager.requestAlwaysAuthorization()
        locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    }
    
    //MARK: UIImagePickerControllerDelegate
    
    func imagePickerController(picker: UIImagePickerController, didFinishPickingImage image: UIImage, editingInfo: [String : AnyObject]?) {
        
        //TODO: SALVAR IMAGEM
        checkin = Checkin.newEntity(context!)
        checkin!.picture = image
        locationManager.requestLocation()
        
        picker.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func imagePickerControllerDidCancel(picker: UIImagePickerController) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    //MARK: CLLocationManagerDelegate
    
    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        let loca = locations.last
        
        if let myLoc = loca {
            
            let checkins: Array<Checkin> = Checkin.select("latitude", byValue: myLoc.coordinate.latitude, orderBy: nil, ascending: nil, onContext: self.context!)
            let exists = checkins.contains({ (c) -> Bool in
                return c.longitude == myLoc.coordinate.longitude
            })
            
            if !exists {
                checkin!.latitude = myLoc.coordinate.latitude
                checkin!.longitude = myLoc.coordinate.longitude
                appDelegate.saveContext()
            } else {
                checkin = nil
            }
            
            print(checkin);
        }
    }
    
    func locationManager(manager: CLLocationManager, didFailWithError error: NSError) {
        print(error.description)
    }
    
    //MARK: LifeCycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.initContextFromAppDelegate()
        self.initLocationManager()
    }
}